packages/editor/src/components/post-switch-to-draft-button/index.js

```diff
@@ -18,37 +22,39 @@ function PostSwitchToDraftButton( {
 	onClick,
 } ) {
 	const isMobileViewport = useViewportMatch( 'small', '<' );
+	const [ showConfirmDialog, setShowConfirmDialog ] = useState( false );
 
 	if ( ! isPublished && ! isScheduled ) {
 		return null;
 	}
 
-	const onSwitch = () => {
-		let alertMessage;
-		if ( isPublished ) {
-			alertMessage = __(
-				'Are you sure you want to unpublish this post?'
-			);
-		} else if ( isScheduled ) {
-			alertMessage = __(
-				'Are you sure you want to unschedule this post?'
-			);
-		}
-		// eslint-disable-next-line no-alert
-		if ( window.confirm( alertMessage ) ) {
-			onClick();
-		}
-	};
+	let alertMessage;
+	if ( isPublished ) {
+		alertMessage = __( 'Are you sure you want to unpublish this post?' );
+	} else if ( isScheduled ) {
+		alertMessage = __( 'Are you sure you want to unschedule this post?' );
+	}
 
 	return (
-		<Button
-			className="editor-post-switch-to-draft"
-			onClick={ onSwitch }
-			disabled={ isSaving }
-			variant="tertiary"
-		>
-			{ isMobileViewport ? __( 'Draft' ) : __( 'Switch to draft' ) }
-		</Button>
+		<>
+			<Button
+				className="editor-post-switch-to-draft"
+				onClick={ () => {
+					setShowConfirmDialog( true );
+				} }
+				disabled={ isSaving }
+				variant="tertiary"
+			>
+				{ isMobileViewport ? __( 'Draft' ) : __( 'Switch to draft' ) }
+			</Button>
+			<ConfirmDialog
+				isOpen={ showConfirmDialog }
+				onConfirm={ onClick }
```

Comment:
Should we also close the dialog when it's confirmed? This seems to be the case with the current implementation (native confirm). E.g.

```js
const handleConfirm = useCallback( () => {
  setShowConfirmDialog( false );
  onClick();
}, [ onClick ]);

// ...

return (
  <>
    {/* ... */}
    <ConfirmDialog
      isOpen={ showConfirmDialog }
      onConfirm={ handleConfirm }
      {/* ... */}
    />

    {/* ... */}
  </>
);
```

Reply:
Happy to change that, though I would like to understand the benefit a little better. In my tests, the dialog does close when confirmed, and when I inspect the next render of the `DeleteMenu` component, `showConfirmDialog` is once again `false`.

If both of those things are happening by default, what do we gain by explicitly closing as you've suggested?

(To be clear, I'm asking from genuine curiosity and a desire to learn, not because I disagree with the recommendation 😊 )

Reply:
> In my tests, the dialog does close when confirmed, and when I inspect the next render of the DeleteMenu component, showConfirmDialog is once again false.

I thought that the dialog doesn't self-close on confirm when the `isOpen` prop is set (see [here](https://github.com/WordPress/gutenberg/blob/59dce7546f86abece759066769d8c60b139cca31/packages/components/src/confirm-dialog/component.tsx#L43-L51)) — that's why I suggested to close it explicitly.

Curious to see then if that snippet of code that I highlighted is not working as expected?

Reply:
Ah, I see your point, thank you for clarifying. The `ConfirmDialog` README also supports that interpretation when in Controlled mode (i.e. if `isOpen` is set):
> You'll want to update the state that controls isOpen by updating it from the onCancel and onConfirm callbacks.

So it really sounds like I should need to explicitly close the dialog on confirm, but I'm not actually needing to for some reason. @fullofcaffeine does this sound like a bug, or is this expected due to a nuance I'm not picking up on?

Reply:
> @fullofcaffeine does this sound like a bug, or is this expected due to a nuance I'm not picking up on?

As far as I can remember, it shouldn't close itself if `isOpen` is passed, so it does sound like a bug 🤔 I'll take note and have a look if no one beats me to it.

Reply:
Could the reason be the same as explained in https://github.com/WordPress/gutenberg/pull/37535#discussion_r784291648 ?

In any case, I believe that we should still call `setShowConfirmDialog( false )` here

Reply:
Sounds good, I'll update the PR either later today or early on Monday!

Reply:
Done in f6fcb32. I didn't apply a `useCallback` because it triggered an error due to the conditional when placed after the early return declared by `if ( ! isPublished && ! isScheduled ) {...}`

Reply:
Thank you!

> I didn't apply a useCallback because it triggered an error due to the conditional when placed after the early return declared by if ( ! isPublished && ! isScheduled ) {...}

That's because hooks, including `useCallback`, can't run conditionally. We can leave the code as-is, but FYI an easy way to get around that issue would have been to declare the callback before that `if` statement :)